### ISSUE TITLE
DTensor support for fused qkv matmul

### ISSUE THREAD

**Issue by HDCharles** [pytorch/pytorch#140069]
### 🚀 The feature, motivation and pitch

For transformer architecture (for example https://github.com/pytorch-labs/gpt-fast/blob/main/model.py#L195-L211) it tends to be most performant to merge the qkv matrices together. If you try to shard this concatenated tensor then the subsequent SDPA op won't be shared correctly since you need each column of q sharded with the corresponding columns of  k and v [q1,k1,v1,...], but by default the sharding will be [q1, q2, q3...] When not using DTensor this is relatively easy to get to work: https://github.com/pytorch-labs/gpt-fast/blob/main/tp.py#L73

but for DTensor the way to enable this is really unclear. is there a way to handle this type of operation with DTensor parallelization or should we just stick to normal tensor parallel support and figure out how to get it to work with our APIs?

This is currently blocking tensor parallel support in torchAO so i wanted to centralize discussion to a single location.

### Alternatives

don't use DTensor for tensor parallel

**Comment by ad8e** [comment]
Depends on DTensor Strided Sharding: 
https://github.com/pytorch/pytorch/issues/88838#issuecomment-2081838245
https://github.com/pytorch/pytorch/issues/129627

**Comment by gnadathur** [comment]
cc: @XilunWu

**Comment by XilunWu** [comment]
@HDCharles sorry for catching this issue late. As @ad8e said, this can be done via Strided Sharding feature in DTensor. I'll try the TP code you shared in DTensor and see if there's any implementation gap. Thanks for submitting the feature request!

**Comment by yzhangcs** [comment]
@XilunWu Hey, wondering any progress on this feature? Thank you!

**Comment by XilunWu** [comment]
Hi @yzhangcs sorry currently we haven't made a plan to support this but I'll notify if we have more updates.

**Comment by tianyu-l** [comment]
@yzhangcs 
How is this related to https://github.com/pytorch/torchtitan/issues/827?
From your description over there it looks existing mechanism in DTensor / TP can already support your fused FFN layer.

For fused QKV, indeed you might need strided sharding as a way to express things.
The expression itself (`_StridedShard`) is already available in pytorch: https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/placement_types.py#L375

It is likely that for your use case, some op support related to `_StridedShard` is missing, but you can try adding by yourself, even if we don't have bandwidth to do so.

**Comment by wanchaol** [comment]
Sorry I think this might be because I mentioned strided sharding could support fused QKV before, which might led the discussions to strided sharding. After looking into this more, I think strided sharding and fused QKV sharding are two orthogonal problems. 

Strided sharding only describes the sharding order when sharding **one tensor** on multiple device mesh dimensions.

Fused QKV sharding is different, it is trying to shard a combined linear (one big tensor) but as if it shards there linears separately on one device mesh dimension. So essentially it tries to **shard three tensor into one big sharded tensor**. I think this is not a regular sharded tensor but one might want to treat it as a regular sharded tensor during runtime. More over the three tensors might have different shapes which need to be traced manually. 

Given this is not a regular sharded tensor, we should not try to force it into a regular one during sharding initialization, instead we should treat it faithfully as shard three tensors to one big sharded tensor, and just treat it as a single big DTensor during runtime.

We can easily implement the fused QKV feature in the TP layer, i.e. sth like below should work:

```
wq, wk, wv = linear.weight.split(fused_qkv_splits)
# distribute there
sharded_wq = distribute_tensor(wq, ...).to_local()
...
sharded_fused_qkv = torch.cat([sharded_wq, sharded_wk..])
dt = DTensor.from_local(sharded_fused_qkv, ..)
linear.register_paramter("weight", dt)
```

cc @HDCharles @ad8e

**Comment by ad8e** [comment]
My understanding of the proposed scheme is: the shape of `self.weight` is `(TP dimension, qkv for one shard fused together, input)`, where dim=0 is the TP dimension. The TP dimension can't be changed, since that would cause the heads to rearrange. If the first two dimensions were merged, producing the total shape `(qkv output size, qkv input size)`, the TP-shape would still be baked into the weight matrix.
The TP-handling mechanism would then be outside of DTensor and DCP; we would rearrange heads with surgery.
This seems ok. We need the same logic with regular torch tensors, so nothing will change.

**Comment by wanchaol** [comment]
@ad8e Yep that is right! Sharding really need to align between pretraining -> finetune or inference stage, i.e. if pretrain use fused qkv sharding, then finetune/inference requires no change. But if pretrain use separate qkv sharding, there needs to be a surgery converting the sharding, as fused qkv trying to treat the whole qkv as one weight, while non-fused one keeping as three weights separately.  From checkpointable states prospective they are different
